doc/HACKING/design/01.00-lib-overview.md

```diff
@@ -0,0 +1,171 @@
+modules sorted from lowest to highest level by running
+`./scripts/maint/practracker/includes.py --toposort`.
+
+As of this writing, the library modules are (from lowest to highest
+level):
+
+   * `lib/cc` -- Macros for managing the C compiler and
+     language. Includes macros for improving compatibility and clarity
+     across different C compilers.
+
+   * `lib/version` -- Holds the current version of Tor.
+
+   * `lib/testsupport` -- Helpers for making test-only code and test
+     mocking support.
+
+   * `lib/defs` -- Lowest-level constants used in many places across the
+     code.
+
+   * `lib/subsys` -- Types used for declaring a "subsystem". A subsystem
+     is a module with support for initialization, shutdown,
+     configuration, and so on.
+
+   * `lib/conf` -- Types and macros used for declaring configuration
+     options.
+
+   * `lib/arch` -- Compatibility functions and macros for handling
+     differences in CPU architecture.
+
+   * `lib/err` -- Lowest-level error handling code: responsible for
+     generating stack traces, handling raw assertion failures, and
+     otherwise reporting problems that might not be safe to report
+     via the regular logging module.
+
+   * `lib/malloc` -- Wrappers and utilities for memory management.
+
+   * `lib/intmath` -- Utilities for integer mathematics.
+
+   * `lib/fdio` -- Utilities and compatibility code for reading and
+      writing data on file descriptors (and on sockets, for platforms
+      where a socket is not a kind of fd).
+
+   * `lib/lock` -- Compatibility code for declaring and using locks.
+      Lower-level than the rest of the threading code.
+
+   * `lib/ctime` -- Constant-time implementations for data comparison
+     and table lookup, used to avoid timing side-channels from standard
+     implementations of memcmp() and so on.
+
+   * `lib/string` -- Low-level compatibility wrappers and utility
+     functions for string manipulation.
+
+   * `lib/wallclock` -- Compatibility and utility functions for
+     inspecting and manipulating the current (UTC) time.
+
+   * `lib/osinfo` -- Functions for inspecting the version and
+     capabilities of the operating system.
+
+   * `lib/smartlist_core` -- The bare-bones pieces of our dynamic array
+     ("smartlist") implementation. There are higher-level pieces, but
+     these ones are used by (and therefore cannot use) the logging code.
+
+   * `lib/log` -- Implements the logging system used by all higher-level
+     Tor code.  You can think of this as the logical "midpoint" of the
+     library code: much of the higher-level code is higher-level
+     _because_ it uses the logging module, and much of the lower-level
+     code is specifically written to avoid having to log, because the
+     logging module depends on it.
+
+   * `lib/container` -- General purpose containers, including dynamic arrays,
+     hashtables, bit arrays, weak-reference-like "handles", bloom
+     filters, and a bit more.
```

Comment:
Would it make sense here to mention with dynamic arrays that we have given them the slightly confusing name of "smartlist"?